crates/noirc_evaluator/src/ssa/acir_gen.rs

```diff
@@ -177,7 +177,50 @@ impl Acir {
                 InternalVar::from(v)
             }
             Operation::Call { .. } => unreachable!("call instruction should have been inlined"),
-            Operation::Return(_) => todo!(), //return from main
+            Operation::Return(node_ids) => {
+                // XXX: When we return a node_id that was created from
+                // the UnitType, there is a witness associated with it
+                // Ideally no witnesses are created for such types.
+
+                // This can only ever be called in the main context.
+                // In all other context's, the return operation is transformed.
+
+                for node_id in node_ids {
+                    // An array produces a single node_id
+                    // We therefore need to check if the node_id is referring to an array
+                    // and deference to get the elements
+                    let objects = match Memory::deref(ctx, *node_id) {
+                        Some(a) => {
+                            let array = &ctx.mem[a];
+                            self.load_array(array, false, evaluator)
+                        }
+                        None => vec![self.substitute(*node_id, evaluator, ctx)],
+                    };
+
+                    for mut object in objects {
+                        // TODO(Open issue): We want to semantically specify that
+                        // TODO you cannot return constant values as public outputs.
+                        // TODO: in that case, you should just have a constant.
+                        // TODO: For now, we will support this usecase.
+                        let witness = if object.expression.is_const() {
+                            evaluator.create_intermediate_variable(object.expression)
+                        } else {
+                            object.generate_witness(evaluator)
+                        };
+
+                        // Before pushing to the public inputs, we need to check that
+                        // it was not a private ABI input
+                        if evaluator.is_private_abi_input(witness) {
+                            return Err(RuntimeErrorKind::Spanless(String::from(
+                                "we do not allow private ABI inputs to be returned as public outputs",
+                            )));
+                        }
+                        evaluator.public_inputs.push(witness);
```

Comment:
we should not have duplicates in this list, and should use a public_outputs list for storing the returned values

Reply:
I mentioned it above -- this will be added when we add `public_outputs` into ACIR